Finish stretchy character support and add bbox support

dpvc opened the pull request:
This PR finishes the stretchy character support (by adding the multi-character support).  It also adds a font data object and uses that to hold the information about delimiters, character bounding boxes, and TeX font parameters for the font, so the TeX font parameters have been moved from where they were.

This also implements bbox computations for TextNodes and mo nodes (which should make the bonding boxes correct for mrow as well, but mfrac still needs work).

This PR is targeted to the stretchy-chars branch, so the you don't see all those original files as well.  If you want, I can merge this into stretchy-chars and you can review them as a whole (but I was trying to do incremental updates to keep the reviews small).  Since this does make changes to stretchy-chars, it might be easier not to have to look at the originals but only final versions at this point.  Let me know how you want to handle it.

dpvc commented:
Just realized I forgot to tslint and add some of the jsdoc comments, so hold off on this until I do that.  Sorry!

dpvc commented:
OK, I've linted and added the missing jsdoc comments.  So this should be ready to review.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     */
+    public createVariants(variants: string[][]) {
+        for (const variant of variants) {
+            this.createVariant.apply(this, variant);
```
Is there a reason why you are using .apply here?

dpvc replied:
Because variant is an array that contains the `name`, `inherit`, and `link` values for the `createVariant()` call, so they need to be broken out.  I suppose you could do

```
this.createVariant(variant[0], variant[1], variant[2]);
``` 
or
```
this.createVariant(...variant);
```
if you prefer.  (I admit that I don't always think of the spread operator; It is hard to unlearn past paradigms.)

dpvc replied:
It turns out Typescript doesn't like the `this.createVariant(...variant)` form, so I used the other more verbose one.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     */
+    constructor() {
+        let CLASS = (this.constructor as typeof FontData);
+        this.params = Object.assign({}, CLASS.defaultParams);
```
Use spread: 
this.params = {...Class.defaultParams};
(At least I believe that was the syntax.

dpvc replied:
Yes, that's right.  I'll change it.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     */
+    public defineChars(name: string, chars: CharMap) {
+        let variant = this.variant[name];
+        Object.assign(variant.chars, chars);
```
dito

dpvc replied:
Again, using spread would destroy the prototype chain that is part of `variant.chars`.  Also, since `variant.chars` may not be empty, you would have to use something like
```
variant.chars = {...variant.chars, ...chars};
```
which would mean you are copying and recreating the object every time `defineChars()` is called.  Adding the new characters into the existing object is more efficient, and is required in order to keep the inheritance intact.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+        let variant = this.variant[name];
+        Object.assign(variant.chars, chars);
+        for (const link of variant.linked) {
+            Object.assign(link, chars);
```
dito

dpvc replied:
ditto

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     * @param{DelimiterMap} delims  The delimiters to define
+     */
+    public defineDelimiters(delims: DelimiterMap) {
+        Object.assign(this.delimiters, delims);
```
dito

dpvc replied:
While there is no inheritance in this case, the `delimiters` may not be empty, and using spread would require copying and recreating the object.  There is no need for that.  This is what `Object.assign()` is for.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     */
+    public getSizeVariant(n: number, i: number) {
+        if (this.delimiters[n].variants) {
+            i = this.delimiters[n].variants[i];
```
This can be undefined. Is that a problem?

dpvc replied:
Note that `getSizeVariant()` is only called when the character is known to be stretchy, i.e., `this.delimiters[n]` is known to exist, so isn't undefined.  For a properly constructed `DelimiterData`, the `variants` array should be the same length as the `sizes` array, but you are correct that there is no guarantee of that.

One could, I suppose, use
```
i = this.delimiters[n].variants[i] || 0;
```
but I think I'd rather get an error when the `undefined` is used as a number in order to find out that the data are not correct rather than hiding it and getting a wrong glyph.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+        if (this.delimiters[n].variants) {
+            i = this.delimiters[n].variants[i];
+        }
+        return this.sizeVariants[i];
```
dito

dpvc replied:
Same here as above.  The `sizeVariants` array should be as large as the largest `sizes` array in any of the delimiters, but again, there is no guarantee that this is true.  One could do
```
return this.sizeVariants[i] || 'normal';
```
I suppose, but again, that is hiding an inconsistency in the font data that should be corrected, not hidden.

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+
+    /*
+     * @param{number} n  The delimiter character number whose data is desired
+     * @return{DelimiterData}  The data for that delimiter (or null)
```
or undefined not null

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+    /*
+     * @param{string} name  The variant whose character data is being querried
+     * @param{number} n     The unicode number for the character to be found
+     * @return{CharData}    The data for the given character (or null)
```
dito

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+
+    /*
+     * @param{string} name   The name of the variant whose data is to be obtained
+     * @return{VariantData}  The data for the requested variant (or null)
```
dito

zorkow commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     * @return{CharData}    The data for the given character (or null)
+     */
+    public getChar(name: string, n: number) {
+        return this.variant[name].chars[n];
```
Case that variant[name] does not exist missing.

dpvc replied:
Shouldn't happen, and if it does, it should produce an error (as it will here) so that the inconsistency in the data can be corrected.  This is a function that will be called for every character displayed, so any unnecessary overhead should be avoided.

zorkow commented:
I found some documentation on my proposed changes:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html
under "spread and rest", as well as here:
https://github.com/tc39/proposal-object-rest-spread/blob/master/Spread.md

dpvc commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+    public createVariant(name: string, inherit: string = null, link: string = null) {
+        let variant = {
+            linked: [] as CharMap[],
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
```
This is not a clone, it creates an object having a prototype of the given object.  This is a lookup chain.  This allows a variant to have its own data (for for example for bold) but fall back on another variant (e.g., normal).  That way, the bold variant doesn't have to repeat all the data in the normal variant.

It is important to have inheritance of variants in this way, since, for example, one can do

```
<mstyle mathvariant="fraktur">
  <mi>x</mi>
  <mo>&gt;</mo>
  <mi>y</mi>
</mstyle>
```
so all three inner elements get `mathvariant="fraktur'`.  But there is no glyph for a plus sign in the MathJax Fraktur font, so it should default back to the normal version.  The prototype chain is what makes that work.

dpvc commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
+        };
+        if (link && this.variant[link]) {
+            Object.assign(variant.chars, this.variant[link].chars);
```
It is a merge, and the properties are the character data arrays (the mapping from unicode number to CharData array).

The reason for this extra link is that for a `mathvariant` like `bold-italic`, you want to inherit from both the `bold` and `italic` variants, but the prototype chain can only inherit from one.  So for `bold-italic`, we make an object that has a prototype consisting of a *copy* of the `bold` data, and add the `italic` data as the prototype chain.  (Since this *is* a copy, we keep a record of this link so that if `bold` is changed later, we update this copy.  That is not needed for the prototype chain, since the prototypes are the actual objects, not copies.)  We then use this bold-plus-italic object as the prototype chain for the `bold-italic` object

That means that `bold-italic` will first look in its own object for specifically bold-italic glyphs that are defined there, then in the copy of the bold glyphs (only its top level is copied, not is prototype chain), and then the specifically italic glyphs, and then the prototype chain for italics, which is the normal glyphs.  Effectively, this means `bold-italic` looks for bold-italic, then bold, then italic, then normal glyphs in order to find the given character.

Something like `bold-script` will link the `script` font and use `bold` as its prototype.  So `bold-script` will look for specifically bold-script glyphs, then `script` glyphs, then `bold` glyphs, then `normal` ones.

This is set up to reduce the amount of duplication in the lists, as well as the amount of copying that must be done when new glyphs are added to variants that are in the default variants (e.g., normal, bold, italic).

Because `variant.chars` is an object that may have a prototype already, we can't use spread to merge the linked chars.  E.g., 

```
variant.chars = {...this.variant[link].chars};
```
would lose the prototype to the inherited characters.

zorkow replied:
Thanks! This is a great explanation, and I believe should be kept for posterity.
Could you please make a comment out of \S 2+3?

dpvc commented on mathjax3-ts/output/chtml/Wrappers/TextNode.ts:
```diff
@@ -51,13 +51,24 @@ export class CHTMLTextNode extends CHTMLWrapper {
+        const variant = this.parent.variant;
+        let bbox = this.bbox;
+        if (variant === '-explicitFont') {
+            // FIXME:  measure this using DOM, if possible
```
This is the location where the code to measure the characters using the DOM will go (or be linked in). The comment is there to indicate that there is more coming in the future.  If it would make you happier, I can add a call to something like `this.getDomBBox()` at this point, and make an empty `getDomBBox()` method.

zorkow replied:
Not necessary.

dpvc commented on mathjax3-ts/output/chtml/FontData.ts:
```diff
@@ -0,0 +1,328 @@
+     *   that includes a copy of the linked variant, and then the inherit variant chain.
+     *   The linked copy is updated automatically when the link variant is modified.
+     *   (The idea is to be able to have something like bold-italic inherit from both
+     *   bold and intalic by having the prototype chain include a copy of bold plus
```
I don't see it.  Can you be more specific?

dpvc commented on mathjax3-ts/output/chtml/Wrappers/mo.ts:
```diff
@@ -65,44 +36,112 @@ export class CHTMLmo extends CHTMLWrapper {
+     */
+    protected stretchHTML(chtml: HTMLElement, symmetric: boolean) {
+        const c = this.getText().charCodeAt(0);
+        const C = this.font.getDelimiter(c).stretch;
```
OK, I changed `C` to `delim`.

dpvc commented:
I have addressed all the issues either by making the change, or explaining the reason for the original code (except for one typo that I couldn't find).  Can you take another look?

zorkow commented:
I approve. But somehow my approve button is gone.

zorkow commented:
It's probably the merge conflict. Anyway, some of your comments to my review should probably be kept for posterity. They are very good!

dpvc commented:
I've added the comments that you asked for into `FontData.js`.  Let me know if you want any of the others as well.

I've fixed the merge conflict, but I don't see an approval button anymore either, so I'm just going to merge it.
